rust/src/ffi/base64.rs

```diff
@@ -26,6 +27,132 @@ pub enum Base64ReturnCode {
     SC_BASE64_OVERFLOW,
 }
 
+#[repr(C)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum Base64Mode {
+    /* If the following strings were to be passed to the decoder with RFC2045 mode,
+     * the results would be as follows. See the unittest B64TestVectorsRFC2045 in
+     * src/util-base64.c
+     *
+     * BASE64("") = ""
+     * BASE64("f") = "Zg=="
+     * BASE64("fo") = "Zm8="
+     * BASE64("foo") = "Zm9v"
+     * BASE64("foob") = "Zm9vYg=="
+     * BASE64("fooba") = "Zm9vYmE="
+     * BASE64("foobar") = "Zm9vYmFy"
+     * BASE64("foobar") = "Zm 9v Ym Fy"   <-- Notice how the spaces are ignored
+     * BASE64("foobar") = "Zm$9vYm.Fy"    # According to RFC 2045, All line breaks or *other
+     * characters* not found in base64 alphabet must be ignored by decoding software
+     * */
+    Base64ModeRFC2045 = 0, /* SPs are allowed during transfer but must be skipped by Decoder */
+    Base64ModeStrict,
+    /* If the following strings were to be passed to the decoder with RFC4648 mode,
+     * the results would be as follows. See the unittest B64TestVectorsRFC4648 in
+     * src/util-base64.c
+     *
+     * BASE64("") = ""
+     * BASE64("f") = "Zg=="
+     * BASE64("fo") = "Zm8="
+     * BASE64("foo") = "Zm9v"
+     * BASE64("foob") = "Zm9vYg=="
+     * BASE64("fooba") = "Zm9vYmE="
+     * BASE64("foobar") = "Zm9vYmFy"
+     * BASE64("f") = "Zm 9v Ym Fy"   <-- Notice how the processing stops once space is encountered
+     * BASE64("f") = "Zm$9vYm.Fy"    <-- Notice how the processing stops once an invalid char is
+     * encountered
+     * */
+    Base64ModeRFC4648, /* reject the encoded data if it contains characters outside the base alphabet */
+}
+
+#[repr(C)]
+pub struct Base64Decoded {
+    decoded: *mut u8,
+    decoded_len: u32,
+}
+
+impl Default for Base64Decoded {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Base64Decoded {
+    pub fn new() -> Base64Decoded {
+        Base64Decoded {
+            decoded: std::ptr::null_mut(),
+            decoded_len: 0,
+        }
+    }
+}
+
+impl Drop for Base64Decoded {
+    fn drop(&mut self) {
+        unsafe {
+            if !self.decoded.is_null() {
+                let _ = Vec::from_raw_parts(
+                    self.decoded,
+                    self.decoded_len as usize,
+                    self.decoded_len as usize,
+                );
+            }
+        }
+    }
+}
+
+/// Base64 decode a buffer.
+///
+/// This method exposes the Rust base64 decoder to C and should not be called from
+/// Rust code.
+///
+/// It allows decoding in the modes described by ``Base64Mode`` enum.
+#[no_mangle]
+pub unsafe extern "C" fn Base64Decode(
+    input: *const u8, len: u32, max_decoded: u32, mode: Base64Mode,
```

Comment:
what I expected here was a way to pass an existing buffer in, that would be filled by this function

Now we still allocate things

Reply:
There is no longer a static buffer on the C side anymore. And, the vector allocated now has only the required size of the decoded string and not the length of the encoded string (which was the length of the static array).
1. Could you please tell why is it still better to pass a static buffer here to be filled in?
2. Do you have an example for that?

Thank you!

Reply:
The stack allocation doesn't exist anymore as this patch removes it :) I argue it should stay and be used instead.

Reply:
`rs_to_hex` is an example

Reply:
Thank you. Attempting today.